include/core/shape.h

```diff
@@ -0,0 +1,594 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019 - 2019 by the Lethe authors
+ *
+ * This file is part of the Lethe library
+ *
+ * The Lethe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------*/
+
+#ifndef lethe_shape_h
+#define lethe_shape_h
+
+#include <deal.II/base/auto_derivative_function.h>
+#include <deal.II/base/function.h>
+
+#if (DEAL_II_VERSION_MAJOR < 10 && DEAL_II_VERSION_MINOR < 4)
+#else
+#  include <deal.II/base/function_signed_distance.h>
+#endif
+
+
+#include <deal.II/physics/transformations.h>
+
+using namespace dealii;
+
+/**
+ * @brief A base class used to represent geometrical entities. Its main uses
+ * are to return signed distance and its gradient. It inherits
+ * AutoDerivativeFunction so that it can evaluate gradients even when they are
+ * not implemented analytically.
+ *
+ * @tparam dim An integer that denotes the dimension of the space in which
+ * the flow is solved
+ */
+template <int dim>
+class Shape : public AutoDerivativeFunction<dim>
+{
+public:
+  /**
+   * @brief A general constructor for the Shapes
+   */
+  Shape(double radius)
+    : AutoDerivativeFunction<dim>(1e-8)
+  {
+    this->effective_radius = radius;
+  }
+
+  virtual ~Shape()
+  {}
+
+  /**
+   * @brief Return the evaluation of the signed distance function of this solid
+   * at the given point p
+   * Most levelset functions implemented come from Inigo Quilez:
+   * iquilezles.org/articles/distfunctions
+   */
+  virtual double
+  value(const Point<dim> & p,
+        const unsigned int component = 0) const override = 0;
+
+  /**
+   * @brief Return a pointer to a copy of the Shape
+   */
+  virtual std::shared_ptr<Shape<dim>>
+  static_copy() const = 0;
+
+  /**
+   * @brief
+   * Return the volume displaced by the solid
+   *
+   */
+  virtual double
+  volume(const double fluid_density)
+  {
+    StandardExceptions::ExcNotImplemented();
+    return 1.;
+  }
+
+  /**
+   * @brief
+   * Most value functions assume that the particle position is at the origin and
+   * that the shape is aligned with one of the main axes. This function moves
+   * the evaluation point so that the value function can be kept simple.
+   *
+   * Returns the centered and aligned point to evaluate properly the levelset
+   * afterwards
+   */
+  Point<dim>
+  real_to_centered(const Point<dim> &evaluation_pt) const
```

Comment:
I think a more meaningful name would help, as well.

Reply:
Is align_and_center a better choice?